### PR TITLE
[Dynamic Partition] Support for automatically drop partitions

### PR DESCRIPTION
**Pull request by WingsGo**
For #3080

### CODE REVIEW

File: fe/src/test/java/org/apache/doris/persist/ModifyDynamicPartitionInfoTest.java [hunk 1]
```diff
@@ -49,6 +49,7 @@ public void testNormal() throws IOException {
         HashMap<String, String> properties = new HashMap<>();
         properties.put(DynamicPartitionProperty.ENABLE, "true");
         properties.put(DynamicPartitionProperty.TIME_UNIT, "day");
+        properties.put(DynamicPartitionProperty.START, "3");
```
Comment (kangkaisen) [inline fe/src/test/java/org/apache/doris/persist/ModifyDynamicPartitionInfoTest.java]: -3?

File: fe/src/test/java/org/apache/doris/catalog/TablePropertyTest.java [hunk 2]
```diff
@@ -49,6 +49,7 @@ public void testNormal() throws IOException {
         HashMap<String, String> properties = new HashMap<>();
         properties.put(DynamicPartitionProperty.ENABLE, "true");
         properties.put(DynamicPartitionProperty.TIME_UNIT, "day");
+        properties.put(DynamicPartitionProperty.START, "3");
```
Comment (kangkaisen) [inline fe/src/test/java/org/apache/doris/catalog/TablePropertyTest.java]: -3

File: fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java [hunk 3]
```diff
@@ -293,6 +294,50 @@ public void testMissTimeUnit(@Injectable SystemInfoService systemInfoService,
         catalog.createTable(stmt);
     }
 
+    @Test
+    public void testMissSTART(@Injectable SystemInfoService systemInfoService,
```
Comment (kangkaisen) [inline fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java]: You could refer to `QueryPlanTest`, and create real table.

File: fe/src/main/java/org/apache/doris/service/Untitled-2.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# 1 Master + 2 Follower
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/service/Untitled-2.md]: Why add this file?

File: fe/src/main/java/org/apache/doris/common/Config.java [hunk 5]
```diff
@@ -981,7 +981,7 @@
     /*
      * Decide how often to check dynamic partition
      */
-    @ConfField(mutable = true, masterOnly = true)
+    @ConfField(mutable = false, masterOnly = true)
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/common/Config.java]:
```suggestion
    @ConfField
```

File: fe/src/main/java/org/apache/doris/service/Untitled-2.md (hunk 4 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/service/Untitled-2.md]: my fault…I will remove it

File: fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java (hunk 3 above)
Comment (WingsGo) [reply fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java]: ok, i will change it.

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 6]
```diff
@@ -136,12 +245,15 @@ private void dynamicAddPartition() {
...
             }
-            String tableName;
-            ArrayList<AddPartitionClause> addPartitionClauses = new ArrayList<>();
+
             db.readLock();
+            ArrayList<AddPartitionClause> addPartitionClauses;
+            ArrayList<DropPartitionClause> dropPartitionClauses;
+            OlapTable olapTable = (OlapTable) db.getTable(tableId);
+            String tableName;
+            boolean skipAddPartition = false;
```
[the remark this answers is not in the document]
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]: done

File: fe/src/main/java/org/apache/doris/common/Config.java (hunk 5 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/common/Config.java]: I think maybe we should allow dynamic change the check interval seconds?

File: fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java (hunk 3 above)
Comment (WingsGo) [reply fe/src/test/java/org/apache/doris/catalog/DynamicPartitionTableTest.java]: done

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 7]
```diff
@@ -168,98 +281,62 @@ private void dynamicAddPartition() {
...
-                    List<String> distColumnNames = new ArrayList<>();
-                    for (Column distributionColumn : hashDistributionInfo.getDistributionColumns()) {
-                        distColumnNames.add(distributionColumn.getName());
-                    }
-                    DistributionDesc distributionDesc = new HashDistributionDesc(dynamicPartitionProperty.getBuckets(), distColumnNames);
-
-                    // add partition according to partition desc and distribution desc
-                    addPartitionClauses.add(new AddPartitionClause(rangePartitionDesc, distributionDesc, null, false));
-                }
+                addPartitionClauses = getAddPartitionClause(olapTable, partitionColumn, partitionFormat);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]: No need to get add partition clause is `skipAddPartition` is true.

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 8]
```diff
@@ -121,11 +123,118 @@ public void createOrUpdateRuntimeInfo(String tableName, String key, String value
...
+            RangePartitionInfo info = (RangePartitionInfo) (partitionInfo);
+            boolean isPartitionExists = false;
+            Range<PartitionKey> addPartitionKeyRange;
+            try {
+                PartitionKey lowerBound = PartitionKey.createPartitionKey(Collections.singletonList(lowerValue), Collections.singletonList(partitionColumn));
+                PartitionKey upperBound = PartitionKey.createPartitionKey(Collections.singletonList(upperValue), Collections.singletonList(partitionColumn));
+                addPartitionKeyRange = Range.closedOpen(lowerBound, upperBound);
+            } catch (AnalysisException e) {
+                // keys.size is always equal to column.size, cannot reach this exception
+                LOG.error("Keys size is not equal to column size.");
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]: Do not use `error` level, error means fatal.
And it better to add details in log, such as e.getMessage().

Same suggest to the `getDropPartitionClause()`

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java (hunk 7 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]: done

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java (hunk 8 above)
Comment (WingsGo) [reply fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]: done

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 9]
```diff
@@ -150,7 +150,7 @@ public void createOrUpdateRuntimeInfo(String tableName, String key, String value
                 addPartitionKeyRange = Range.closedOpen(lowerBound, upperBound);
             } catch (AnalysisException e) {
                 // keys.size is always equal to column.size, cannot reach this exception
-                LOG.error("Keys size is not equal to column size.");
+                LOG.warn("Keys size is not equal to column size. Error=" + e.getMessage());
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]:
```suggestion
                LOG.warn("Keys size is not equal to column size. Error={}", e.getMessage());
```

File: fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 10]
```diff
@@ -211,7 +211,7 @@ public void createOrUpdateRuntimeInfo(String tableName, String key, String value
             reservePartitionKeyRange = Range.closedOpen(lowerBound, upperBound);
         } catch (AnalysisException e) {
             // keys.size is always equal to column.size, cannot reach this exception
-            LOG.error("Keys size is not equal to column size.");
+            LOG.warn("Keys size is not equal to column size. Error=" + e.getMessage());
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java]:
```suggestion
            LOG.warn("Keys size is not equal to column size. Error={}", e.getMessage());
```

**Review by morningman — DISMISSED** [review]
LGTM

**Review by kangkaisen — APPROVED** [review]
+1， LGTM

**Pull request merged**